internal/directory_downloader.go

```diff
@@ -0,0 +1,83 @@
+package internal
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+	"github.com/wal-g/wal-g/utility"
+)
+
+type DirectoryDownloader interface {
+	DownloadDirectory(pathToRestore string) error
+}
+
+type DirectoryIsNotEmptyError struct {
+	error
+}
+
+type CommonDirectoryDownloader struct {
+	Folder     storage.Folder
+	BackupName string
+}
+
+func NewCommonDirectoryDownloader(folder storage.Folder, backupName string) DirectoryDownloader {
```

Comment:
I guess it is better to pass the explicit list of tars for download to the downloader constructor. Sometimes we don't need to download all the tars from the folder.

Reply:
If you have explicit list of tars then you can use `ExtractAll` function from package `internal`